docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,185 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
+
+**Property 1**: Only a single message is processed at a time in a subnet. So message execution is sequential, and never parallel.
+
+**Property 2**: Each call (query / update) triggers a message. When an inter-canister call is made using `await`, the code after the call (the callback, highlighted in blue) is executed as a separate message. 
+
+Note that if the code does not `await` the response, the code after the callback (until the next inter-canister call is triggered using `await`) is executed in the same message.
+
+For example, consider the following Motoko code: 
+
+
+![example_highlighted_code](_attachments/example_highlighted_code.png)
+
+The first message that is executed here are lines 2-3, until the inter-canister call is made using the `await` syntax (orange box). The second message executes lines 3-5: when the inter-canister call returns (blue box). We call this part the _callback_ of the inter-canister call. The two messages involved in this example will always be scheduled sequentially. 
+
+**Property 3**: Messages from interleaving calls have no reliable execution ordering.
+
+Let's consider the above example code again, and assume the method `example` is called twice in parallel, the resulting calls being Call 1 and Call 2. The following illustration shows two possible message orderings. On the left, the first call's messages are scheduled first, and only then the second call's messages are executed. On the right, we see another possible message scheduling, where the first messages of each call are executed first. Your code should result in a correct state regardless of the message ordering.
+
+![example_orderings](_attachments/example_orderings.png)
+
+**Property 4**: On a trap / panic, modifications to the canister state for the current message are not applied.
+
+For example, if a trap in the second message (blue box) of the above example occurs, canister state changes resulting from that message, even earlier in the blue box, are discarded. However, note that any state changes from earlier messages and in particular the first message (orange box) have been applied, as that message executed successfully. 
+
+**Property 5**: Inter-canister messaging is not reliable.
+
+Every inter-canister call is guaranteed to receive a response. However, the response does not have to be successful, but can also be a reject response. The reject may come from the called canister, but it may also be generated by the Internet Computer. Such system-generated rejects can occur at any time, and cannot be controlled by canister authors. Thus, it's important that the calling canister handles reject responses as well. Note that a reject response generally guarantees that the message had not been processed by the called canister.
```

Comment:
```suggestion
Every inter-canister call is guaranteed to receive a response. However, the response does not have to be successful, but can also be a reject response. The reject may come from the called canister, but it may also be generated by the Internet Computer. Such system-generated rejects can occur at any time, and cannot be controlled by canister authors. Thus, it's important that the calling canister handles reject responses as well. Note that a reject response generally guarantees that the message had not been processed successfully by the called canister.
```

It could have been processed, but not to completion, right? Or is that unnecessary nitpicking?